Implement getBuildArtifacts tool and optimize getBuild response

dt-atmosic opened the pull request:
This PR implements issue https://github.com/jenkinsci/mcp-server-plugin/issues/32 by adding new tools for artifact management and optimizing the getBuild response size.

### New Tools Added

1. getBuildArtifacts - Returns only the artifacts array from a build
o Takes jobFullName and optional buildNumber parameters
o Significantly reduces payload size compared to full getBuild response
o Returns empty array for jobs without artifacts or non-existent jobs
o Fixed JSON concatenation issue for top-level lists

2. getBuildArtifact - Provides paginated access to individual artifact file contents
o Supports offset/limit parameters for large files
o Returns artifact content as text with pagination metadata
o Includes safety limits to prevent excessive memory usage

### Optimizations

3. Modified getBuild response - Now excludes artifacts field
o Created RunWithoutArtifactsSerializer to filter out artifacts
o Updated JenkinsExportedBeanSerializerModifier to use custom serializer
o Maintains all other build information while reducing response size

### Testing done

4. Comprehensive test coverage
o BuildArtifactsExtensionTest - Integration tests for new tools
o GetBuildWithoutArtifactsTest - Verifies getBuild optimization
o BuildArtifactsExtensionCompileTest - Basic compilation verification

5. Manual test
o Augment Code MCP client w/ `mvn hpi:run` - Found JSON issue with top-level list
o Manual test with Augment Code and production Jenkins server

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests that demonstrate the feature works or the issue is fixed

dt-atmosic committed:
Implement getBuildArtifacts tool and optimize getBuild response

- Add BuildArtifactsExtension with getBuildArtifacts and getBuildArtifact tools
- Create RunWithoutArtifactsSerializer to exclude artifacts from getBuild responses
- Add comprehensive tests for new functionality
- Implement pagination support for large artifact files
- Maintain backward compatibility while reducing payload sizes

Addresses first part of issue #32: separate artifacts retrieval from build metadata

dt-atmosic committed:
Fix compilation issues in tests and serializer

- Replace Java record with regular class for Java 11 compatibility
- Fix regex pattern in RunWithoutArtifactsSerializer with DOTALL flag
- Update test methods to use proper McpSchema.TextContent pattern
- Remove non-existent TestUtils.getTextContent() method calls
- Follow existing test patterns from the codebase

dt-atmosic committed:
Fix objectMapper variable scope issue in test

- Declare ObjectMapper inside lambda scope where it's used
- Resolves compilation error in GetBuildWithoutArtifactsTest.java

dt-atmosic committed:
Test fixes and cleanup

- Use record for BuildArtifactResponse
- Use TreePruner
- spotless:apply

dt-atmosic committed:
Fix JSON responses for getJobs and getBuildArtifacts

dt-atmosic committed:
Update README for getBuildArtifact*

dt-atmosic committed:
Merge remote-tracking branch 'origin/main' into artifacts

dt-atmosic commented:
Working on the code coverage warnings.

dt-atmosic committed:
Improve code coverage for BuildArtifactsExtension

olamy commented on src/main/java/io/jenkins/plugins/mcp/server/extensions/BuildArtifactsExtension.java:
```diff
@@ -0,0 +1,192 @@
+     * - "unchecked": Run.getArtifacts() returns raw List requiring unchecked conversion
+     */
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    @Tool(description = "Get the artifacts for a specific build or the last build of a Jenkins job")
```
will need annotations = @Tool.Annotations(destructiveHint = false) 
as it's `true` per default

dt-atmosic replied:
Done — added `annotations = @Tool.Annotations(destructiveHint = false)` to `getBuildArtifacts`.

olamy commented on src/main/java/io/jenkins/plugins/mcp/server/extensions/BuildArtifactsExtension.java:
```diff
@@ -0,0 +1,192 @@
+                .orElse(List.of());
+    }
+
+    @Tool(
```
will need annotations = @Tool.Annotations(destructiveHint = false) 
as it's `true` per default

dt-atmosic replied:
Done — added `annotations = @Tool.Annotations(destructiveHint = false)` to `getBuildArtifact` as well.

olamy commented on src/main/java/io/jenkins/plugins/mcp/server/extensions/BuildArtifactsExtension.java:
```diff
@@ -0,0 +1,192 @@
+            offset = 0L;
+        }
+        if (limit == null || limit <= 0) {
+            limit = 65536; // 64KB default
```
might be better having a configuration (e.g SystemProperties) as users may want to configure more or less.

dt-atmosic replied:
Done. The default is now read from a system property: `SystemProperties.getInteger("io.jenkins.plugins.mcp.server.extensions.BuildArtifactsExtension.limit.default", 65536)`, so it still defaults to 64KB but can be tuned.

olamy reviewed (changes requested):
Nice PR.
the title says `optimize getBuild response`. is it remove artifacts from the tree?

dt-atmosic committed:
Merge remote-tracking branch 'origin/main' into agent-01KSV8F361553V4EZXC06JZW1B

# Conflicts:
#	README.md
#	src/main/java/io/jenkins/plugins/mcp/server/tool/McpToolWrapper.java
#	src/test/java/io/jenkins/plugins/mcp/server/EndPointTest.java
#	src/test/java/io/jenkins/plugins/mcp/server/extensions/DefaultMcpServerTest.java

dt-atmosic committed:
Address PR review: annotations, configurable limits, binary handling, tree-prune compat

- Add @Tool.Annotations(destructiveHint=false) to getBuildArtifacts/getBuildArtifact

- Make default/max artifact size limits configurable via SystemProperties

- Detect binary artifacts and refuse with a clear message instead of garbled text

- Rework RunWithoutArtifactsSerializer to honor the upstream 'tree' parameter while still always excluding artifacts

- Update tests for the ToolResponse {status,message,result} wrapper and add binary-artifact coverage

dt-atmosic commented on src/main/java/io/jenkins/plugins/mcp/server/extensions/BuildArtifactsExtension.java:
```diff
@@ -0,0 +1,192 @@
+        }
+
+        // Cap the limit to prevent excessive memory usage
+        final int maxLimit = 1048576; // 1MB max
```
Done. The cap is now `SystemProperties.getInteger("io.jenkins.plugins.mcp.server.extensions.BuildArtifactsExtension.limit.max", 1048576)` — defaults to 1MB, configurable.

dt-atmosic commented on src/main/java/io/jenkins/plugins/mcp/server/tool/McpToolWrapper.java:
```diff
@@ -222,13 +222,9 @@ McpSchema.CallToolResult toMcpResult(Object result) {
-            }
+            // Serialize all results the same way - this fixes the JSON concatenation issue
+            // for top-level lists while maintaining proper JSON structure
+            resultBuilder.addTextContent(toJson(result));
```
Agreed. I've merged the latest `main`, which includes the `ToolResponse` refactor from #82. I dropped this local `toMcpResult` change in favor of upstream's implementation, so list/single results are now handled consistently there.

dt-atmosic commented on src/main/java/io/jenkins/plugins/mcp/server/tool/McpToolWrapper.java:
```diff
@@ -222,13 +222,9 @@ McpSchema.CallToolResult toMcpResult(Object result) {
-            } else {
-                resultBuilder.addTextContent(toJson(result));
-            }
+            // Serialize all results the same way - this fixes the JSON concatenation issue
```
Removed — this whole `toMcpResult` change was reverted in favor of upstream's `ToolResponse` (#82), so the comment no longer exists.

dt-atmosic commented on src/main/java/io/jenkins/plugins/mcp/server/extensions/BuildArtifactsExtension.java:
```diff
@@ -0,0 +1,192 @@
+            }
+
+            // Convert to string (assuming text content)
+            String content = new String(buffer, 0, bytesRead, StandardCharsets.UTF_8);
```
Good point. `getBuildArtifact` now detects binary content (NUL-byte sniff over the read buffer) and returns a clear message — *"Artifact appears to be binary and cannot be returned as text: <path>"* — instead of emitting garbled UTF-8. `getBuildArtifacts` still lists every artifact (binary or not) for discovery. Added test coverage in `testGetBuildArtifactBinary`. Let me know if you'd prefer base64 streaming instead.

dt-atmosic commented:
> Nice PR. the title says `optimize getBuild response`. is it remove artifacts from the tree?

Thanks!  Yes, that's correct.

novinxy commented:
Hi, nice PR, just something we need from the MCP 👍
@dt-atmosic isn't it already finished? In my opinion and from the conversation it looks as everything/almost everything is resolved
Is there something I could do to help make it happen ?

dustymabe commented:
Hi. Thanks for working on this!

I notice we now have some filtering for binary artifacts. One use case to consider is people running a `$test_framework` inside of Jenkins to perform tests and the tool generates logs itself (additional logs to the ones output to stdout/stderr). As part of the job run the logs from the tests are uploaded as artifacts in a tar.xz to the build. In this case the tar.xz would be detected as binary and filtered out IIUC. 

Ultimately the use case I'm presenting is something like having an Agent/LLM via jenkins MCP say:

> oh, build `#100` failed and from the Jenkins logs I see test `A234` failed so crack open the tarball and look at the logs for `A234` to investigate further. And oh btw build `#99` didn't fail on the same test, so if you need to compare a successful run to a failed one look at the logs for `A234` from build `#99`.
